libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-						);
+					param_arg_map[i] = nullptr;
+					not_all_mapped = true;
+					m_errorReporter.typeError(
+						_functionCall.location(),
+						"Named argument \"" +
+						*argumentNames[i] +
+						"\" does not match function declaration."
+					);
 				}
+			}
+
+			if (not_all_mapped)
+				return;
+		}
+	}
+
+	// Check for compatible types between arguments and parameters
+	for (size_t i = 0; i < param_arg_map_size; ++i)
+		if (!type(*param_arg_map[i])->isImplicitlyConvertibleTo(*parameterTypes[i]))
+		{
+			solAssert(!!param_arg_map[i], "unmapped parameter");
+
+			string msg =
+				"Invalid type for argument in function call. "
+				"Invalid implicit conversion from " +
+				type(*param_arg_map[i])->toString() +
+				" to " +
+				parameterTypes[i]->toString() +
+				" requested.";
+			if (
+				_functionType->kind() == FunctionType::Kind::BareCall ||
+				_functionType->kind() == FunctionType::Kind::BareCallCode ||
+				_functionType->kind() == FunctionType::Kind::BareDelegateCall ||
+				_functionType->kind() == FunctionType::Kind::BareStaticCall
+			)
+				msg +=
+					" This function requires a single bytes argument."
+					" If all your arguments are value types, you can"
+					" use abi.encode(...) to properly generate it.";
+			else if (
+				_functionType->kind() == FunctionType::Kind::KECCAK256 ||
+				_functionType->kind() == FunctionType::Kind::SHA256 ||
+				_functionType->kind() == FunctionType::Kind::RIPEMD160
+			)
+				msg +=
+					" This function requires a single bytes argument."
+					" Use abi.encodePacked(...) to obtain the pre-0.5.0"
+					" behaviour or abi.encode(...) to use ABI encoding.";
+			m_errorReporter.typeError(param_arg_map[i]->location(), msg);
 		}
+}
+
+bool TypeChecker::visit(FunctionCall const& _functionCall)
+{
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+	bool argumentsArePure = true;
+
+	// We need to check arguments' type first as they will be needed for overload resolution.
+	for (ASTPointer<Expression const> const& argument: arguments)
+	{
+		argument->accept(*this);
+		if (!argument->annotation().isPure)
+			argumentsArePure = false;
+	}
+
+	// For positional calls only, store argument types
+	if (_functionCall.names().empty())
+	{
+		shared_ptr<TypePointers> argumentTypes = make_shared<TypePointers>();
+		for (ASTPointer<Expression const> const& argument: arguments)
+			argumentTypes->push_back(type(*argument));
+		_functionCall.expression().annotation().argumentTypes = move(argumentTypes);
 	}
 
-	if (returnTypes.size() == 1)
-		_functionCall.annotation().type = returnTypes.front();
+	_functionCall.expression().accept(*this);
+
+	TypePointer const expressionType = type(_functionCall.expression());
+	TypePointer actualType;
+
+	// Determine type for type conversion or struct construction expressions
+	if (expressionType->category() == Type::Category::TypeType)
+		if (auto const* const typeType = dynamic_cast<TypeType const*>(expressionType.get()))
```

Comment:
Should not be conditional but rather throw if conversion fails:
```suggestion
		actualType = dynamic_cast<TypeType const&>(*expressionType)).actualType();
```

Reply:
This was set early on and kept for type conversions to be passed in later on (null for vanilla functions).
I've moved it down into the block that handles type conversions/struct ctors - it will throw there if null.

The separate type check method for type conversions now extracts the expression type from the function call node itself rather than relying on the visit function to pass it in - I believe this is cleaner.